internal/core/services/events_forwarder/events_forwarder_service.go

```diff
@@ -24,60 +24,169 @@ package events_forwarder
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"strconv"
+
+	"github.com/topfreegames/maestro/internal/core/entities"
+	"github.com/topfreegames/maestro/internal/core/entities/game_room"
+
+	"github.com/topfreegames/maestro/internal/core/entities/forwarder"
 
 	"github.com/topfreegames/maestro/internal/core/entities/events"
 
-	"github.com/topfreegames/maestro/internal/core/entities/game_room"
 	"github.com/topfreegames/maestro/internal/core/ports"
 	"github.com/topfreegames/maestro/internal/core/services/interfaces"
 	"go.uber.org/zap"
 )
 
 type EventsForwarderService struct {
-	eventsForwarder ports.EventsForwarder
-	logger          *zap.Logger
+	eventsForwarder  ports.EventsForwarder
+	logger           *zap.Logger
+	schedulerStorage ports.SchedulerStorage
+	instanceStorage  ports.GameRoomInstanceStorage
+	schedulerCache   ports.SchedulerCache
 }
 
-func NewEventsForwarderService(eventsForwarder ports.EventsForwarder) interfaces.EventsService {
+func NewEventsForwarderService(
+	eventsForwarder ports.EventsForwarder,
+	schedulerStorage ports.SchedulerStorage,
+	instanceStorage ports.GameRoomInstanceStorage,
+	schedulerCache ports.SchedulerCache,
+) interfaces.EventsService {
 	return &EventsForwarderService{
 		eventsForwarder,
 		zap.L().With(zap.String("service", "rooms_api")),
+		schedulerStorage,
+		instanceStorage,
+		schedulerCache,
 	}
 }
 
 func (es *EventsForwarderService) ProduceEvent(ctx context.Context, event *events.Event) error {
-	room := &game_room.GameRoom{}
-	instance := &game_room.Instance{}
-
-	switch event.Name {
-	case events.RoomEvent:
-		return es.forwardRoomEvent(ctx, room, instance, map[string]interface{}{}, "")
-	case events.PlayerEvent:
-		return es.forwardPlayerEvent(ctx, room, map[string]interface{}{}, "")
+	var eventType string
+	if _, ok := event.Attributes["eventType"]; ok {
+		eventType = event.Attributes["eventType"].(string)
+	} else {
+		return errors.New("eventAttributes must contain key \"eventType\"")
 	}
-	return nil
 
+	var scheduler *entities.Scheduler
+	var err error
+	scheduler, _ = es.schedulerCache.GetScheduler(ctx, event.SchedulerID)
```

Comment:
to avoid code smell (Bloaters), I think you could create an internal func to get scheduler and there you put logic to find scheduler, get from cache etc...

then just replace the lines between 76 to 86 to
scheduler := es.getScheduler(ctx, event.SchedulerID)

Reply:
A lot better for sure! Thanks!!